File: 315. Count of Smaller Numbers After Self.java
class Solution {
    public List<Integer> countSmaller(int[] nums) {
        List<Integer> answer = new ArrayList<>();
        int max = Integer.MIN_VALUE;
        int min = Integer.MAX_VALUE;
        for (int i : nums) {
            max = Math.max(max, i+100000);
            min = Math.min(min, i+100000);
        }
        SegmentTree segmentTree = new SegmentTree(min, max);
        for (int i = nums.length-1; i >= 0; i--) {
            answer.add(segmentTree.find(nums[i]+100000));
            segmentTree.add(nums[i]+100000);
        }
        Collections.reverse(answer);
        return answer;
    }
}

class SegmentTree {
    int min;
    int max;
    int count;
    SegmentTree left;
    SegmentTree right;
    
    SegmentTree(int min, int max) {
        this.min = min;
        this.max = max;
    }

    int find(int x) {
        if (x > this.max) return count;
        int mid = min + (max-min)/2;
        if (x <= mid) {
            if (left == null) return 0;
            return this.left.find(x);
        } else {
            int answer = 0;
            if (left != null) answer += left.find(x);
            if (right != null) answer += right.find(x);
            return answer;
        }
    }

    void add(int x) {
        if (x > max || x < min) return;
        this.count++;
        if (max <= min) return;
        int mid = min + (max-min)/2;
        if (x <= mid) {
            if (left == null) left = new SegmentTree(min, mid);
            left.add(x);
        } else {
            if (right == null) right = new SegmentTree(mid+1, max);
            right.add(x);
        }
        
    }

}

File: go/315. Count of Smaller Numbers After Self.go
func countSmaller(nums []int) []int {
    st := &segmentTree{0, 20000, 0, nil, nil}
    answer := []int{}
    for i := len(nums)-1; i >= 0; i-- {
        answer = append(answer, st.find(nums[i]+10000))
        st.add(nums[i]+10000)
    }
    for i := 0; i < len(answer)/2; i++ {
        answer[i], answer[len(answer)-1-i] = answer[len(answer)-1-i], answer[i]
    }
    return answer
}

type segmentTree struct {
	min int
	max int
	count int
	left *segmentTree
	right *segmentTree
}

func (s *segmentTree) find(x int) int {
	if s.max < x {
		return s.count
	}
	mid := (s.min + s.max)/2
	if mid+1 >= x {
		if s.left == nil { return 0 }
		return s.left.find(x)
	} else {
		answer := 0
		if s.left != nil { answer += s.left.find(x) }
		if s.right != nil { answer += s.right.find(x) }
		return answer
	}
}

func (s *segmentTree) add(x int) {
	if s.min > x || s.max < x { return }
	s.count++
	if s.max <= s.min { return }
	mid := (s.min + s.max) / 2
	if x <= mid {
		if s.left == nil { s.left = &segmentTree{s.min, mid, 0, nil, nil} }
		s.left.add(x)
	} else {
		if s.right == nil { s.right = &segmentTree{mid+1, s.max, 0, nil, nil} }
		s.right.add(x)
	}
}